docs/core/testing/unit-testing-best-practices.md

```diff
@@ -0,0 +1,331 @@
+
+### Executable documentation
+It may not always be obvious what a particular method does or how it behaves given a certain input. You may ask yourself: How does this method behave if I pass it a blank string? Null?
+
+When you have a suite of well-named unit tests, each test should be able to clearly explain the expected output for a given input. In addition, it should be able to verify that it actually works.
+
+### Less coupled code
+When code is tightly coupled, it can be difficult to unit test. Without creating unit tests for the code that you're writing, coupling may be less apparent.
+
+Writing tests for your code will naturally decouple your code, because it would be near impossible to test otherwise.
+
+## Characteristics of a good unit test
+- **Fast**. It is not uncommon for mature projects to have thousands of unit tests. Unit tests should take very little time to run. Milliseconds.
+- **Isolated**. Unit tests are standalone, can be run in isolation, and have no dependencies on any outside factors such as a file system or database.
+- **Repeatable**. Running a unit test should be consistent with its results, that is, it always returns the same result if you do not change anything in between runs.
+- **Self-Checking**. The test should be able to automatically detect if it passed or failed without any human interaction.
+- **Timely**. A unit test should not take a disproportionally long time to write compared to the code being tested. If you find testing the code taking a large amount of time compared to writing the code, consider a design that is more testable.
+
+## Lets speak the same language
+The term *mock* is unfortunately very misused when talking about testing. The following defines the most common types of *fakes* when writing unit tests:
+
+*Fake* - A fake is a generic term which can be used to describe either a stub or a mock object. Whether it is a stub or a mock depends on the context in which it's used. So in other words, a fake can be a stub or a mock.
+
+*Mock* - A mock object is a fake object in the system that decides whether or not a unit test has passed or failed. A mock starts out as a Fake until it is asserted against.
+
+*Stub* - A stub is a controllable replacement for an existing dependency (or collaborator) in the system. By using a stub, you can test your code without dealing with the dependency directly. By default, a fake starts out as a stub.
+
+Consider the following code snippet:
+
+```csharp
+var mockOrder = new MockOrder();
+var purchase = new Purchase(mockOrder);
+
+purchase.ValidateOrders();
+
+Assert.True(purchase.CanBeShipped);
+```
+
+This would be an example of Mock being used improperly. In this case, it is a stub. We're just passing in the Order as a means to be able to instantiate `Purchase` (the system under test). The name `MockOrder` is also very misleading because again, the order is not a mock.
```

Comment:
instead of "Mock being used improperly" consider "stub being referred to as a mock". (The code doesn't misuse a mock, it's using the wrong name).